tests/unit/test_grub.py

```diff
@@ -0,0 +1,488 @@
+    """Test validation error and it's properties."""
+    exp_key, exp_message = "test", "test message"
+
+    error = grub.ValidationError(exp_key, exp_message)
+    with pytest.raises(ValueError):
+        raise error
+
+    assert error.key == exp_key
+    assert error.message == exp_message
+    assert str(error) == exp_message
+
+
+@pytest.mark.parametrize(
+    "output, exp_result",
+    [
+        (mock.MagicMock(return_value=b"lxd"), True),
+        (mock.MagicMock(side_effect=subprocess.CalledProcessError(1, [])), False),
+    ],
+)
+def test_is_container(output, exp_result):
+    """Test helper function to validate if machine is container."""
+    with mock.patch("subprocess.check_output", new=output) as mock_check_output:
+        assert grub.is_container() == exp_result
+        mock_check_output.assert_called_once_with(
+            ["/usr/bin/systemd-detect-virt", "--container"], stderr=subprocess.STDOUT
+        )
+
+
+class BaseTestGrubLib(unittest.TestCase):
+    def setUp(self) -> None:
+        tmp_dir = tempfile.TemporaryDirectory()
+        self.tmp_dir = Path(tmp_dir.name)
+        self.addCleanup(tmp_dir.cleanup)
+
+        # configured paths
+        grub.GRUB_DIRECTORY = self.tmp_dir
+
+        # change logger
+        mocked_logger = mock.patch.object(grub, "logger")
+        self.logger = mocked_logger.start()
+        self.addCleanup(mocked_logger.stop)
+
+
+class TestGrubUtils(BaseTestGrubLib):
+    def test_parse_config(self):
+        """Test parsing example GRUB config with skipping duplicated key."""
+        stream = io.StringIO(GRUB_CONFIG_EXAMPLE)
+        result = grub._parse_config(stream)
+
+        self.assertEqual(result, EXP_GRUB_CONFIG)
+
+    def test_parse_config_with_duplicity(self):
+        """Test parsing example GRUB config with skipping duplicated key."""
+        raw_config = (
+            GRUB_CONFIG_EXAMPLE + 'GRUB_CMDLINE_LINUX_DEFAULT="$GRUB_CMDLINE_LINUX_DEFAULT pti=on"'
+        )
+        stream = io.StringIO(raw_config)
+        result = grub._parse_config(stream)
+
+        self.logger.warning.assert_called_once_with(
+            "key %s is duplicated in config", "GRUB_CMDLINE_LINUX_DEFAULT"
+        )
+        self.assertEqual(
+            result["GRUB_CMDLINE_LINUX_DEFAULT"], "'\"$GRUB_CMDLINE_LINUX_DEFAULT pti=on\"'"
+        )
+
+    def test_load_config_not_exists(self):
+        """Test load config from file which does not exist."""
+        path = self.tmp_dir / "test_load_config"
+
+        with pytest.raises(FileNotFoundError):
+            grub._load_config(path)
+
+    @mock.patch.object(grub, "_parse_config")
+    def test_load_config(self, mock_parse_config):
+        """Test load config from file."""
+        exp_config = {"test": "valid"}
+        mock_parse_config.return_value = exp_config
+        path = self.tmp_dir / "test_load_config"
+        path.touch()  # create file
+
+        with mock.patch.object(grub, "open", mock.mock_open()) as mock_open:
```

Comment:
I understand wanting to mock the subprocess calls, but why do we need to mock the open calls? Especially seeing the test is creating temp directories -- can't we do real open/read on real files in the tests? (Local FS operations are fast and reliable.) Or am I missing something here.

Reply:
There is no point to actually reading it, since `_parse_config` is mocked and I want to check that open was called correctly.